pkg/lib/_internal/dnf5daemon.ts

```diff
@@ -297,28 +338,296 @@ export class Dnf5DaemonManager implements PackageManager {
         }
 
-        const client = dbus_client();
-        const subscription = client.subscribe({}, signal_emitted);
-        let session: string | null = null;
+        await this.with_session(async (session) => {
+            try {
+                await call(session, "org.rpm.dnf.v0.rpm.Rpm", "install", [data.missing_names, {}]);
+                const [, resolve_result] = await call(session, "org.rpm.dnf.v0.Goal", "resolve", [{}]);
 
-        try {
-            session = await open_session();
-            await call(session, "org.rpm.dnf.v0.rpm.Rpm", "install", [data.missing_names, {}]);
-            const [, resolve_result] = await call(session, "org.rpm.dnf.v0.Goal", "resolve", [{}]);
+                if (resolve_result !== 0) {
+                    const [problem] = await call(session, "org.rpm.dnf.v0.Goal", "get_transaction_problems_string", []);
+                    throw new ResolveError(`Resolving install failed with result=${resolve_result} ${problem}`);
+                }
+                await call(session, "org.rpm.dnf.v0.Goal", "do_transaction", [{}]);
+            } catch (err) {
+                console.warn("install error", err);
+            }
+        }, signal_emitted);
+    }
 
+    async refresh(_force: boolean, _progress_cb?: ProgressCB): Promise<void> {
+        await this.with_session(async (session) => {
+            // refresh dnf5daemon state
+            await call(session, "org.rpm.dnf.v0.Base", "read_all_repos", []);
+            const [, resolve_result] = await call(session, "org.rpm.dnf.v0.Goal", "resolve", [{}]) as [unknown[], number];
             if (resolve_result !== 0) {
                 const [problem] = await call(session, "org.rpm.dnf.v0.Goal", "get_transaction_problems_string", []);
-                throw new ResolveError(`Resolving install failed with result=${resolve_result} ${problem}`);
+                throw new ResolveError(`Resolving read_all_repos failed with result=${resolve_result} - ${problem}`);
             }
+
             await call(session, "org.rpm.dnf.v0.Goal", "do_transaction", [{}]);
-        } catch (err) {
-            console.warn("install error", err);
-        } finally {
-            if (session)
-                await close_session(session);
+        });
+    }
+
+    async is_installed(pkgnames: string[]): Promise<boolean> {
+        const uninstalled = new Set(pkgnames);
+
+        await this.with_session(async (session) => {
+            const package_attrs = ["name", "is_installed"];
+
+            const [results] = await call(session, "org.rpm.dnf.v0.rpm.Rpm", "list", [
+                {
+                    package_attrs: { t: 'as', v: package_attrs },
+                    scope: { t: 's', v: "all" },
+                    patterns: { t: 'as', v: pkgnames }
+                }
+            ]) as ListPackage[][];
+
+            for (const pkg of results) {
+                if (pkg.is_installed.v) {
+                    uninstalled.delete(pkg.name.v);
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22611-09ca2e27-20251209-111826-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)